### PR TITLE
Rewrite coerce_plan_expr_for_schema to fix union type coercion

### LINKED ISSUE #4837: SQL statement (`UNION` + `EXCEPT`) causes panic
**Describe the bug**
`datafusion 15.0.0` panics for specific SQL-query:

```sql
(   
    SELECT * FROM table_1
    EXCEPT
    SELECT * FROM table_2
)  
UNION ALL
(   
    SELECT * FROM table_2
    EXCEPT
    SELECT * FROM table_1
) 
```

**To Reproduce**

```rust
    let table_1_path = "...";
    let table_2_path = "...";

    let ctx = SessionContext::new();

    ctx.register_csv("table_1", table_1_path, CsvReadOptions::default())
        .await?;
    ctx.register_csv("table_2", table_2_path, CsvReadOptions::default())
        .await?;

    let data_frame = ctx.sql("(SELECT * FROM table_1 EXCEPT SELECT * FROM table_2) UNION ALL (SELECT * FROM table_2 EXCEPT SELECT * FROM table_1)").await?;

    data_frame.show().await?;
```

It will panic:
`thread 'main' panicked at 'index out of bounds: the len is 2 but the index is 2'` (`len` & `index` depends on number of columns in file)

**Expected behavior**
It should works correctly or return an error, but not to panic.

### PR DESCRIPTION
**Pull request by ygf11**
# Which issue does this PR close?

Closes #4837.

# Rationale for this change
The original `coerce_plan_expr_for_schema` method use the `input.expressions()` as source of type coercion, which is not correct, because it maybe return other expressions, like join will return its predicates.

# What changes are included in this PR?

The source of union type coercion is from input schema or projection exprs now. 

# Are these changes tested?

Yes.

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/tests/sqllogictests/test_files/union.slt
Comment (ygf11) [inline datafusion/core/tests/sqllogictests/test_files/union.slt:19]: `sqllogictest` is very easy to use, I like it 👍.

**Comment by alamb** [comment]
Thanks @ygf11  -- I hope to review this PR more carefully tomorrow

**Comment by liukun4515** [comment]
I want to review this pr which is about type coercion.

File: datafusion/expr/src/expr_rewriter.rs
Comment (alamb) [inline datafusion/expr/src/expr_rewriter.rs:525]: I was somewhat confused about this at first because the name of this function 

it turns out it is only called for coercing union schemas: https://github.com/search?q=repo%3Aapache%2Farrow-datafusion%20coerce_plan_expr_for_schema&type=code

👍

File: datafusion/expr/src/expr_rewriter.rs
Comment (alamb) [inline datafusion/expr/src/expr_rewriter.rs:529]:
```suggestion
        // special case Projection to avoid adding multiple projections
        LogicalPlan::Projection(Projection { expr, input, .. }) => {
```

I didn't understand the need to special case `LogicalPlan::projection` -- I think it would help to explain the rationale 


For anyone who is curious I tried removing this case and got errors like the following:

```



---- sorted_union_with_different_types_and_group_by stdout ----
thread 'sorted_union_with_different_types_and_group_by' panicked at 'assertion failed: `(left == right)`
  left: `"Sort: a ASC NULLS LAST\n  Union\n    Projection: CAST(x.a AS Float64) AS a\n      Projection: x.a\n        Aggregate: groupBy=[[x.a]], aggr=[[]]\n          SubqueryAlias: x\n            Projection: Int64(1) AS a\n              EmptyRelation\n    Projection: x.a\n      Aggregate: groupBy=[[x.a]], aggr=[[]]\n        SubqueryAlias: x\n          Projection: Float64(1.1) AS a\n            EmptyRelation"`,
 right: `"Sort: a ASC NULLS LAST\n  Union\n    Projection: CAST(x.a AS Float64) AS a\n      Aggregate: groupBy=[[x.a]], aggr=[[]]\n        SubqueryAlias: x\n          Projection: Int64(1) AS a\n            EmptyRelation\n    Projection: x.a\n      Aggregate: groupBy=[[x.a]], aggr=[[]]\n        SubqueryAlias: x\n          Projection: Float64(1.1) AS a\n            EmptyRelation"`', datafusion/sql/tests/integration_test.rs:2175:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panicking.rs:65:14
   2: core::panicking::assert_failed_inner
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panicking.rs:246:17
   3: core::panicking::assert_failed
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panicking.rs:203:5
   4: integration_test::quick_test
             at ./tests/integration_test.rs:2175:5
   5: integration_test::sorted_union_with_different_types_and_group_by
             at ./tests/integration_test.rs:1714:5
   6: integration_test::sorted_union_with_different_types_and_group_by::{{closure}}
             at ./tests/integration_test.rs:1700:1
   7: core::ops::function::FnOnce::call_once
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/ops/function.rs:251:5
   8: core::ops::function::FnOnce::call_once
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/ops/function.rs:251:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Which are due to another projection having been addeed

File: datafusion/expr/src/expr_rewriter.rs
Comment (alamb) [inline datafusion/expr/src/expr_rewriter.rs:569]: I think you can write this more concisely like:

```suggestion
                    Expr::Alias(e, alias) => {
                        Ok(e.clone()
                            .cast_to(new_type, src_schema)?
                            .alias(alias))
                    }
```

File: datafusion/expr/src/expr_rewriter.rs
Comment (alamb) [inline datafusion/expr/src/expr_rewriter.rs:555]:
```suggestion
    exprs: Vec<Expr>,
```

Since the caller owns the Exprs already, I think you can avoid the clones in this function  by passing in the Vec directly

**Review by alamb — APPROVED** [review]
I reviewed this PR carefully both code and tests and it makes sense to me. Thank you @ygf11 

I will wait for @liukun4515 's review prior to merging

File: datafusion/expr/src/expr_rewriter.rs
Comment (ygf11) [reply datafusion/expr/src/expr_rewriter.rs:530]: Yes, it is to avoid adding multiple projections.

File: datafusion/core/tests/sql/union.rs
Comment (ygf11) [inline datafusion/core/tests/sql/union.rs:174]: This plan is changed after merging master, cause #4849 enable the `projection push down` of `DISTINCT`.

Before merge:
```
"      Distinct: [name:UInt8;N]",
"        Projection: t1.name [name:UInt8;N]",
"          TableScan: t1 projection=[name] [name:UInt8;N]",
```

After merge: 
```
"      Distinct: [name:UInt8;N]",
"        TableScan: t1 projection=[name] [name:UInt8;N]",
```

File: datafusion/core/tests/sql/union.rs
Comment (ygf11) [inline datafusion/core/tests/sql/union.rs:216]: Same as the above test.

File: datafusion/core/tests/sql/union.rs
Comment (ygf11) [inline datafusion/core/tests/sql/union.rs:166]: This plan is changed after merging master, cause #4849 enable the `projection push down` of `DISTINCT`.

Before merge:
```
"      Distinct: [name:UInt8;N]",
"        Projection: t1.name [name:UInt8;N]",
"          TableScan: t1 projection=[name] [name:UInt8;N]",
```

After merge: 
```
"      Distinct: [name:UInt8;N]",
"        TableScan: t1 projection=[name] [name:UInt8;N]",
```

File: datafusion/core/tests/sql/union.rs
Comment (ygf11) [inline datafusion/core/tests/sql/union.rs:216]: Same as the above test.

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Thanks again @ygf11

**Pull request merged**
